### PR TITLE
[MODEL-8952] Adds a file to report fit statistics

### PR DESCRIPTION
**Pull request by scottp-ml**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Adds the ability to generate a yaml file containing information gathered while fitting a model.  At the moment this includes fit memory usage, original dataset filesize, and the predict server memory usage.  This will be expanded in the future to include other metrics as well.  

## Rationale
Memory usage is critical to properly allocate resources for tasks within DataRobot.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 1]
```diff
@@ -1147,6 +1153,23 @@ def _get_requirements_lines(reqs_file_path):
...
+            pred_mem_usage: Memory footprint of running the check for prediction side effects
+        """
+        report_information = {
+            "fit_memory_usage": fit_mem_usage,
+            "prediction_memory_usage": pred_mem_usage,
+            "input_filesize": os.path.getsize(self.options.input) / 1e6,
+        }
+        # in run_predict the code dir is set to the output and output is set to /dev/null
+        output_path = Path(self.options.code_dir) / "fit_runtime_data.yaml"
+        yaml.dump(report_information, open(output_path, "w"))
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/drum.py:1171]: nit: json is faster and IMO easier to work wtih

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 2]
```diff
@@ -1147,6 +1153,23 @@ def _get_requirements_lines(reqs_file_path):
...
+            fit_mem_usage: Memory footprint of running the fit job
+            pred_mem_usage: Memory footprint of running the check for prediction side effects
+        """
+        report_information = {
+            "fit_memory_usage": fit_mem_usage,
+            "prediction_memory_usage": pred_mem_usage,
+            "input_filesize": os.path.getsize(self.options.input) / 1e6,
+        }
+        # in run_predict the code dir is set to the output and output is set to /dev/null
+        output_path = Path(self.options.code_dir) / "fit_runtime_data.yaml"
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/drum.py:1170]: Lets make this filename a global var somewhere

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 3]
```diff
@@ -1147,6 +1153,23 @@ def _get_requirements_lines(reqs_file_path):
...
+        Saves information related to running a fit pipeline.  All data is reported in Mb
+
+        Parameters:
+            fit_mem_usage: Memory footprint of running the fit job
+            pred_mem_usage: Memory footprint of running the check for prediction side effects
+        """
+        report_information = {
+            "fit_memory_usage": fit_mem_usage,
+            "prediction_memory_usage": pred_mem_usage,
+            "input_filesize": os.path.getsize(self.options.input) / 1e6,
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/drum.py:1167]: I think this is getting the size of the directory, not the contents in the directory. Needs to be recursive

File: tests/drum/test_fit.py [hunk 4]
```diff
@@ -915,3 +915,38 @@ def test_fit_fails_code_dir_is_output_dir(self, resources, tmp_path):
...
+            cmd += " --enable-fit-metadata"
+        _, stdout, _ = _exec_shell_cmd(
+            cmd, "Failed in {} command line! {}".format(ArgumentsOptions.MAIN_COMMAND, cmd)
+        )
+        if skip_predict:
+            assert "Prediction started" not in stdout
+            assert "predictions can be made on the fit model" not in stdout
+            assert not os.path.exists(output / "fit_runtime_data.yaml")
+        elif output_fit_metadata:
+            print(os.listdir(output))
```
Comment (jmbannon) [inline tests/drum/test_fit.py:951]: remove

File: custom_model_runner/datarobot_drum/drum/args_parser.py [hunk 5]
```diff
@@ -907,6 +917,8 @@ def get_arg_parser():
 
         CMRunnerArgsRegistry._reg_arg_strict_validation(fit_parser, push_parser)
 
+        CMRunnerArgsRegistry._reg_arg_report_fit_predict_metadata(fit_parser, push_parser)
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/args_parser.py:920]: probably needs a test in test_args_parser

File: tests/drum/test_fit.py [hunk 6]
```diff
@@ -915,3 +916,37 @@ def test_fit_fails_code_dir_is_output_dir(self, resources, tmp_path):
...
+            cmd += " --enable-fit-metadata"
+        _, stdout, _ = _exec_shell_cmd(
+            cmd, "Failed in {} command line! {}".format(ArgumentsOptions.MAIN_COMMAND, cmd)
+        )
+        if skip_predict:
+            assert "Prediction started" not in stdout
+            assert "predictions can be made on the fit model" not in stdout
+            assert not os.path.exists(output / FIT_METADATA_FILENAME)
+        elif output_fit_metadata:
+            assert os.path.exists(output / FIT_METADATA_FILENAME)
```
Comment (jmbannon) [inline tests/drum/test_fit.py:952]: Can you add some asserts on the keys and type instances here

**Comment by scottp-ml** [comment]
jarvis check this

**Comment by scottp-ml** [comment]
jarvis test this

**Comment by andrey-hura** [comment]
jarvis test this

**Comment by andrey-hura** [comment]
jarvis please test this

**Comment by andrey-hura** [comment]
jarvis please test this

**Comment by andrey-hura** [comment]
jenkins test black

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 7]
```diff
@@ -672,17 +678,18 @@ def run_test_predict(self):
...
+            else:
+
                 CMRunTests(
-                    self.options, self.run_mode, self.target_type, self.schema_validator
+                    self.options, self.target_type, self.schema_validator
                 ).check_prediction_side_effects()
-            except DrumPredException as e:
-                self.logger.warning(e)
+        except DrumPredException as e:
+            self.logger.warning(e)
```
Comment (drenfr01) [inline custom_model_runner/datarobot_drum/drum/drum.py:692]: Out of curiosity, why are you removing self.run_mode from `CMRunTests` above? Is it an unused param?

File: custom_model_runner/datarobot_drum/drum/perf_testing.py [hunk 8]
```diff
@@ -465,7 +465,7 @@ def _run_test_case(self, tc, results):
         bar.finish()
 
         # TODO: even if prediction request fail we should try and get server stats
-        response = requests.get(self._url_server_address + self._stats_endpoint)
+        response = self.check_server_stats()
```
Comment (drenfr01) [inline custom_model_runner/datarobot_drum/drum/perf_testing.py:469]: Is this a new function from another PR?

**Review by drenfr01 — APPROVED** [review]
LGTM, just a few questions but nothing to block the PR

**Review by rvorobii — APPROVED** [review]

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 7 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/drum.py:692]: Yes, it was unused.

File: custom_model_runner/datarobot_drum/drum/perf_testing.py (hunk 8 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/perf_testing.py:469]: unintended change, I thought I was going to extract a method here and thought I had reverted everything.

**Review by jmbannon — APPROVED** [review]

**Pull request merged**